metrics-datadog/src/main/java/org/coursera/metrics/datadog/RegexMetricFilter.java

```diff
@@ -0,0 +1,46 @@
+package org.coursera.metrics.datadog;
+
+import com.codahale.metrics.Metric;
+import com.codahale.metrics.MetricFilter;
+import com.google.common.collect.ImmutableSet;
+
+import java.util.regex.Pattern;
+
+public class RegexMetricFilter implements MetricFilter {
```

Comment:
I moved the class here to honor the concerns from Nick's first comment in this PR. In theory a Regex-based MetricFilter isn't specific to Dropwizard itself, but is potentially relevant to anything that reports to Datadog (the DatadogReporter class and its test class actually reference other MetricFilters). Even though it's only used by our Dropwizard ReportFactory at the moment, it should reasonably be available to other classes for future use.

As for the tests, I modified them to be more focused on the different behavior that results when "useRegexFilters" is toggled on or off. That still addresses the code in the DatadogReporterFactory class, so I've kept it in the dropwizard-metrics-datadog package. In my first iteration of the code I was more concerned with the use of regular expressions changing the partial-match behavior of our filters, and the tests reflected that. Now that the use of regular expressions is more explicit I no longer think that it's necessarily worth testing. I didn't include any tests for the RegexMetricFilter itself since I couldn't think of any that were more meaningful than simply testing that java.util.regex.Pattern behaves as expected. If you have any good suggestions for tests I'd be happy to add them.